storm-core/src/jvm/org/apache/storm/executor/bolt/BoltOutputCollectorImpl.java

```diff
@@ -0,0 +1,168 @@
+import org.apache.storm.daemon.Task;
+import org.apache.storm.hooks.info.BoltAckInfo;
+import org.apache.storm.hooks.info.BoltFailInfo;
+import org.apache.storm.stats.BoltExecutorStats;
+import org.apache.storm.task.IOutputCollector;
+import org.apache.storm.tuple.MessageId;
+import org.apache.storm.tuple.Tuple;
+import org.apache.storm.tuple.TupleImpl;
+import org.apache.storm.tuple.Values;
+import org.apache.storm.utils.Time;
+import org.apache.storm.utils.Utils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BoltOutputCollectorImpl implements IOutputCollector {
+
+    private static final Logger LOG = LoggerFactory.getLogger(BoltOutputCollectorImpl.class);
+
+    private final BoltExecutor executor;
+    private final Task taskData;
+    private final int taskId;
+    private final Random random;
+    private final boolean isEventLoggers;
+    private final boolean isDebug;
+
+    public BoltOutputCollectorImpl(BoltExecutor executor, Task taskData, int taskId, Random random,
+                                   boolean isEventLoggers, boolean isDebug) {
+        this.executor = executor;
+        this.taskData = taskData;
+        this.taskId = taskId;
+        this.random = random;
+        this.isEventLoggers = isEventLoggers;
+        this.isDebug = isDebug;
+    }
+
+    public List<Integer> emit(String streamId, Collection<Tuple> anchors, List<Object> tuple) {
+        return boltEmit(streamId, anchors, tuple, null);
+    }
+
+    @Override
+    public void emitDirect(int taskId, String streamId, Collection<Tuple> anchors, List<Object> tuple) {
+        boltEmit(streamId, anchors, tuple, taskId);
+    }
+
+    private List<Integer> boltEmit(String streamId, Collection<Tuple> anchors, List<Object> values, Integer targetTaskId) {
+        List<Integer> outTasks;
+        if (targetTaskId != null) {
+            outTasks = taskData.getOutgoingTasks(targetTaskId, streamId, values);
+        } else {
+            outTasks = taskData.getOutgoingTasks(streamId, values);
+        }
+
+        for (Integer t : outTasks) {
+            Map<Long, Long> anchorsToIds = new HashMap<>();
+            if (anchors != null) {
+                for (Tuple a : anchors) {
+                    long edgeId = MessageId.generateId(random);
+                    ((TupleImpl) a).updateAckVal(edgeId);
+                    for (Long root_id : a.getMessageId().getAnchorsToIds().keySet()) {
+                        putXor(anchorsToIds, root_id, edgeId);
+                    }
+                }
+            }
+            MessageId msgId = MessageId.makeId(anchorsToIds);
+            TupleImpl tupleExt = new TupleImpl(executor.getWorkerTopologyContext(), values, taskId, streamId, msgId);
+            executor.getExecutorTransfer().transfer(t, tupleExt);
+        }
+        if (isEventLoggers) {
+            executor.sendToEventLogger(executor, taskData, values, executor.getComponentId(), null, random);
+        }
+        return outTasks;
```

Comment:
`(or out-tasks [])` seems to guarantee at least empty list. Does this guarantee the same?